include/utils.hpp

```diff
@@ -140,6 +173,79 @@ concept vectorizable_v = std::constructible_from<stdx::simd<T>>;
+struct simdize_size<Tup> : simdize_size<std::tuple_element_t<0, Tup>> {
+    static_assert([]<std::size_t... Is>(std::index_sequence<Is...>) {
+        return ((simdize_size<std::tuple_element_t<0, Tup>>::value == simdize_size<std::tuple_element_t<Is, Tup>>::value)
+                && ...);
+    }(std::make_index_sequence<std::tuple_size_v<Tup>>()));
+};
+
+template<typename T>
+inline constexpr std::size_t simdize_size_v = simdize_size<T>::value;
+
+namespace detail {
+/**
+ * Shortcut to determine the stdx::simd specialization with the most efficient ABI tag for the
+ * requested element type T and width N.
+ */
+template<typename T, std::size_t N>
+using deduced_simd = stdx::simd<T, stdx::simd_abi::deduce_t<T, N>>;
+
+template<typename T, std::size_t N>
+struct simdize_impl;
+
+template<vectorizable_v T, std::size_t N>
+requires requires { typename stdx::native_simd<T>; }
+struct simdize_impl<T, N> {
+    using type = deduced_simd<T, N == 0 ? stdx::native_simd<T>::size() : N>;
+};
+
+template<std::size_t N>
+struct simdize_impl<std::tuple<>, N> {
+    using type = std::tuple<>;
+};
+
+template<tuple_like Tup, std::size_t N>
+    requires requires { typename simdize_impl<std::tuple_element_t<0, Tup>, N>::type; }
+struct simdize_impl<Tup, N> {
+    static inline constexpr std::size_t size = N > 0 ? N : []<std::size_t... Is>(std::index_sequence<Is...>) constexpr {
+        return std::max({ simdize_size_v<typename simdize_impl<std::tuple_element_t<Is, Tup>, 0>::type>... });
+    }(std::make_index_sequence<std::tuple_size_v<Tup>>());
+
+    using type = decltype([]<std::size_t... Is>(std::index_sequence<Is...>)
+                                  -> std::tuple<typename simdize_impl<std::tuple_element_t<Is, Tup>, size>::type...> {
+        return {};
+    }(std::make_index_sequence<std::tuple_size_v<Tup>>()));
+};
+} // namespace detail
+
+/**
+ * Meta-function that turns a vectorizable type or a tuple-like (recursively) of vectorizable types
+ * into a stdx::simd or std::tuple (recursively) of stdx::simd. If N is non-zero, N determines the
+ * resulting SIMD width. Otherwise, of all vectorizable types U the maximum
+ * stdx::native_simd<U>::size() determines the resulting SIMD width.
+ */
+template<typename T, std::size_t N = 0>
+using simdize = typename detail::simdize_impl<T, N>::type;
+
+static_assert(std::same_as<simdize<std::tuple<std::tuple<int, double>, short, std::tuple<float>>>,
+                           std::tuple<std::tuple<detail::deduced_simd<int, stdx::native_simd<short>::size()>,
+                                                 detail::deduced_simd<double, stdx::native_simd<short>::size()>>,
+                                      stdx::native_simd<short>,
+                                      std::tuple<detail::deduced_simd<float, stdx::native_simd<short>::size()>>>>);
+
 template<typename A, typename B>
 struct wider_native_simd_size : std::conditional<(stdx::native_simd<A>::size() > stdx::native_simd<B>::size()), A, B> {};
```

Comment:
Aren't these definitions better be declared in the stdlibc++ and vir-simd header files? Do we have to maintain these definitions long-term of each user-level library that uses SIMD?

Reply:
I want to move this stuff up in the dependency chain. I will propose it for standardization one reflection materializes. Until then it cannot live in libstdc++. So vir-simd will be the target. Do you want it in vir-simd before we can merge it? My goal here was to have it mature a little more before releasing it as public API.